### PR TITLE
REF: dont consolidate in is_mixed_type

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Avoid copies, side-effects.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -5471,8 +5471,16 @@ def _consolidate(self, inplace: bool_t = False):
...
-        return self._protect_consolidate(f)
+        if len(self._mgr.blocks) == 1:
+            return False
+
+        if self._mgr.any_extension_types:
+            # Even if they have the same dtype, we cant consolidate them,
+            #  so we pretend this is "mixed'"
+            return True
+
+        dtypes = self.dtypes.unique()
```
Comment (jreback) [inline pandas/core/generic.py:5482]: you can use nunique()

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1547,8 +1547,7 @@ def _setitem_with_indexer(self, indexer, value):
         info_axis = self.obj._info_axis_number
 
         # maybe partial set
-        # _is_mixed_type has the side effect of consolidating in-place
-        take_split_path = self.obj._is_mixed_type
+        take_split_path = len(self.obj._mgr.blocks) > 1
```
Comment (jreback) [inline pandas/core/indexing.py:1550]: worth defining ```__len__``` BlockManager ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py:1550]: i think that would be ambiguous as to whether it was this or `len(mgr.items)`

File: pandas/core/generic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/generic.py:5482]: sure

Comment (jbrockmendel) [reply pandas/core/generic.py]: updated+green

**Review by jreback — APPROVED** [review]

**Pull request merged**